Migrating to CK5: PR2-Initialise CKeditor 5

NishealJ opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
This PR starts the initialization of CKEditor-5 we have divided the migration to CKEditor5 in parts and to know more about it please visit [here](https://docs.google.com/document/d/1V3TaTRnYvrmEw3jp41skj5EHsVcn0SAgSsXdNRDFVjk/edit).

PR #7175 Renames all the CKEditor directives so that they can be distinguished and can be simultaneously used. 

Further, the work is continued in this PR

This PR installs CK5 through npm, initializes basic API for CK5 (getting the view of the editor) and adds CK5 files in the needed places.

This PR creates a flag constant to switch between the CK4 and CK5 in use.

we can use CKEditor 4 keeping the flag as true and once the migration is done we can toggle the flag to false, to start using CK5

currently, the CK5 is initialized with default toolbar to check the view and display issue, in the next PR the directive will be updated to save/discard the data and all the CK4 plugins will be rebuilt for CK5

**Issues/Errors :** 

**Quick Links :** 

Migrating to CK5: PR1-Rename the CK4 directives: visit [here](https://github.com/oppia/oppia/pull/7175)

To follow the steps for CKEditor migration please visit [here](https://docs.google.com/document/d/1V3TaTRnYvrmEw3jp41skj5EHsVcn0SAgSsXdNRDFVjk/edit) 

**Testing Docs :** 

CK PR1 & PR2 [here](https://docs.google.com/document/d/1_fpoq21HDSVcTRTT7dmT2cy3tkskxTI0hx7JK7sM47s/edit?usp=sharing)

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

NishealJ committed:
Update README.md

NishealJ committed:
Update README.md

NishealJ committed:
Update README.md

NishealJ committed:
Update README.md

NishealJ committed:
Update README.md

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Update README.md

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Rename the directive to ck-editor-4

NishealJ committed:
Run CI tests in oppia

NishealJ committed:
Initialize ck5 with ck4

NishealJ committed:
Add the flag

NishealJ committed:
Add code-owners

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

apb7 commented:
Hi @NishealJ, a couple of comments:
1. Please note the failing back-end tests.
2. Please add a proper explanation of what this PR does and if it is stacked on #7175, please mention the commit from where this PR needs to be reviewed.

Thanks!

Commit:
Merge develop

NishealJ commented:
> Hi @NishealJ, a couple of comments:
> 
> 1. Please note the failing back-end tests.
> 2. Please add a proper explanation of what this PR does and if it is stacked on #7175, please mention the commit from where this PR needs to be reviewed.
> 
> Thanks!

Hi @apb7,
Added, was completing the document.

Also, currently the linter test is failing because of the issue `CKEditor-duplicated-modules: Some CKEditor 5 modules are duplicated.` issues triaged in CKEditor-5 [here](https://urlzs.com/YMXT8).
There is no breakage from this error but there is a console log error that's why Travis is failing

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

nithusha21 commented:
Travis is still failing?

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ commented:
Hi @nithusha21,

They are failing for an issue which is already triaged in CKEditor repo [here](https://urlzs.com/YMXT8), they have some solution but as you know this is a stacked PR and currently, in the starting stage of the migration which will be fully migrated in my next milestone it is difficult to apply the suggested solution at the current stage

apb7 commented:
> Hi @nithusha21,
They are failing for an issue which is already triaged in CKEditor repo here, they have some solution but as you know this is a stacked PR and currently, in the starting stage of the migration which will be fully migrated in my next milestone it is difficult to apply the suggested solution at the current stage

@NishealJ, I think we discussed this on chat some time back. Since all of these are logged in the console, the idea was to add them as exceptions in the e2e tests while checking for console errors, at the moment and then eventually remove them when the migration is completed. Also, @nithusha21, do you have any suggestions in this regard? Thanks!

NishealJ committed:
Ignore console errors

NishealJ committed:
Expected by me

NishealJ committed:
Change username

NishealJ commented:
Hi @apb7,
PTAL I've added into the CKE console error to ignore errors Thanks!

NishealJ commented:
Hi @apb7, [here](https://docs.google.com/document/d/1_fpoq21HDSVcTRTT7dmT2cy3tkskxTI0hx7JK7sM47s/edit) is the testing doc for the Ckeditor PR1 & PR2. 

I've tested the CKEditor on two pages Thanks!

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Resolve merge conflicts

NishealJ committed:
Fix merge conflicts

apb7 commented on .github/CODEOWNERS:
```diff
@@ -359,6 +359,7 @@
 
 # Rich text editor team.
 /core/templates/dev/head/components/ck-editor-helpers/ck-editor-4-rte.directive.ts @aks681
+/core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts @apb7 @NishealJ
```
Hmm, I don't think I should be the code owner for this file. It would be better if you replace me with @aks681. (@aks681, does that sound fine?)

NishealJ replied:
Hi @aks681, is that fine?

nithusha21 replied:
Considering he owns ck-editor-4-rte, I think yes.

NishealJ replied:
Done Added @aks681 Thanks!

apb7 commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,98 @@
+          }
+        });
+
+        /**
```
Please use single-line comments here.

NishealJ replied:
Done

apb7 commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,98 @@
+        var blockOverlayRule = ' div(oppia-rte-component-overlay);';
+        // Put all the rules together.
+        var extraAllowedContentRules = componentRule +
+                                       inlineWrapperRule +
```
Try fitting some of these on the same line, until you reach the 80-char limit.

NishealJ replied:
Done Thanks!

apb7 commented on core/templates/dev/head/components/forms/schema-based-editors/schema-based-html-editor.directive.html:
```diff
@@ -1,9 +1,14 @@
 <!-- label-for-focus-target feature is removed since CKEditor has a
 focusManager to manage the focus activity automatically. -->
-<ck-editor-4-rte ng-if="!$ctrl.isDisabled()" ng-model="$ctrl.localValue"
+
```
Remove this line break.

NishealJ replied:
Done Thanks!

nithusha21 replied:
Not done?

apb7 commented on core/tests/protractor_utils/general.js:
```diff
@@ -41,6 +41,14 @@ var checkForConsoleErrors = function(errorsToIgnore) {
       });
     }
+
+    // Currently we are using CKeditor-4 and Ckeditor-5 simultaneously we are
```
Please change this to:
```
    // Since we are in the process of upgrading CKeditor-4 to Ckeditor-5, we are observing
    // consoles errors for duplicated modules (for more details, please refer to https://github.com/ckeditor/ckeditor5/issues/1821)
    // and therefore, we need to filter such logs until we've upgraded to CKEditor-5 completely.
```

Also, I see that the upstream issue is still open (https://github.com/ckeditor/ckeditor5/issues/1821). How do you plan to fix it when the migration is complete?

NishealJ replied:
Hi @apb7, yes the issue is still opened.
They have provided a solution to it but currently, we are using two different CKeditor-4 and CKeditor-5 simultaneously and both are of different build configuration (Classic and Inline). Now we are using CK-5 classic builds, in which all the files are used with plugins which is one of the reasons for this console error. I'll try to reframe the class build editor and try to remove the extra files after the plugins are migrated or I'll try shift use a custom CK build.

apb7 replied:
Okay, sounds good. Thanks for the explanation. Also, I think you haven't yet correctly framed the comments. Please recheck this carefully. Thanks!

NishealJ replied:
Updated PTAL Thansk!

apb7 reviewed (commented):
Hi @NishealJ, left some comments. @nithusha21, can you please take the next pass when these comments have been addressed? Thanks!

nithusha21 commented:
Sure Apurv. @NishealJ can you assign me once you address Apurv's comments?

NishealJ committed:
Resolve comments

NishealJ committed:
Address comments-2

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into ck-0-2

NishealJ commented:
Hi @apb7, Addressed all your comments PTAL Thanks!

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

apb7 commented:
Hi @NishealJ, two e2e tests seem to be failing. PTAL, thanks!

NishealJ committed:
Update expected errors in e2e

NishealJ commented on core/tests/protractor/accessibility.js:
```diff
@@ -49,7 +49,9 @@ describe('Cache Slugs', function() {
-      'http://localhost:9001/build/fail/logo/288x128_logo_white.png'
+      'ckeditor-duplicated-modules',
+      'http://localhost:9001/build/fail/logo/288x128_logo_white.png',
+      'You are using the ngTouch module'
```
Hi @apb7, can we add this warning to in expected errors?

NishealJ replied:
`You are using the ngTouch module` this warning message?

NishealJ committed:
Remove ng-touch message from expected

apb7 commented:
Hi @NishealJ, ptal at this: https://github.com/oppia/oppia/pull/7184#discussion_r306641220. Thanks!

NishealJ commented:
> Hi @NishealJ, ptal at this: [#7184 (comment)](https://github.com/oppia/oppia/pull/7184#discussion_r306641220). Thanks!

Updated the Comments Thanks!

apb7 commented on core/tests/protractor_utils/general.js:
```diff
@@ -41,6 +41,16 @@ var checkForConsoleErrors = function(errorsToIgnore) {
     }
+
+    // Since we are in the process of upgrading CKeditor-4 to Ckeditor-5,
+    // we are observing consoles errors for duplicated modules for more details,
```
I think you're still missing `(` before `for more details...` and `)` after `refer to ...1821`. Please see my original comment https://github.com/oppia/oppia/pull/7184#discussion_r306641220 carefully.

NishealJ replied:
Done PTAL!

apb7 reviewed (commented):
Hi @NishealJ, ptal. Also, one of the e2e tests is still failing. This PR needs to go in soon. Thanks!

apb7 commented:
Hi @nithusha21, I think you can now take a pass over this. @NishealJ has addressed all of my comments except this: https://github.com/oppia/oppia/pull/7184#discussion_r307116367. Thanks!

NishealJ committed:
Add ()

NishealJ commented:
> Hi @nithusha21, I think you can now take a pass over this. @NishealJ has addressed all of my comments except this: [#7184 (comment)](https://github.com/oppia/oppia/pull/7184#discussion_r307116367). Thanks!

Updated it Also e2e tests were failing due to general error Thanks!

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge Develop

nithusha21 commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+          <HTMLElement>(el[0].children[0].children[1]));
+
+        // A RegExp for matching rich text components.
+        var componentRe = (
```
Can we name this something more explicit? maybe componentRegExp or something like that?

NishealJ replied:
Done, updated it to the suggested var name, I've used `componentRe` because we have used it the directives earlier and it was easy to learn and implement

nithusha21 replied:
I don't think this is done?

NishealJ replied:
updated PTAL, Had few changes because of failing test cases.
Also, we don't use it now in much places

nithusha21 commented on core/templates/dev/head/pages/creator-dashboard-page/suggestion-modal-for-creator-view/suggestion-modal-for-creator-view.directive.html:
```diff
@@ -20,6 +20,7 @@ <h3>Review Suggestion</h3>
       </div>
       <div ng-show="suggestionEditorIsShown">
         <ck-editor-4-rte ng-model="suggestionData.newSuggestionHtml"></ck-editor-4-rte>
+        <ck-editor-5-rte ng-model="suggestionData.newSuggestionHtml"></ck-editor-5-rte>
```
Don't you need ng-ifs over here?

NishealJ replied:
Missed now updated PTAL!

nithusha21 replied:
Just curious, how was this missed while testing? It would have displayed 2 RTEs right?

NishealJ replied:
I Think it will display the last-mentioned RTE

nithusha21 commented on core/templates/dev/head/pages/exploration-player-page/templates/exploration-player-suggestion-modal.directive.html:
```diff
@@ -5,6 +5,7 @@ <h3>Suggest a Change</h3>
 <div class="modal-body protractor-test-exploration-suggestion-modal">
   <!-- ng-model needs to bind to a property of an object on the scope (the property cannot sit directly on the scope). -->
   <ck-editor-4-rte ng-show="showEditor" ng-model="suggestionData.suggestionHtml"></ck-editor-4-rte>
+  <ck-editor-5-rte ng-show="showEditor" ng-model="suggestionData.suggestionHtml"></ck-editor-5-rte>
```
Don't you need ng-ifs here and elsewhere?

NishealJ replied:
same here

nithusha21 commented on core/tests/protractor_utils/general.js:
```diff
@@ -41,6 +41,17 @@ var checkForConsoleErrors = function(errorsToIgnore) {
+    // please refer to https://github.com/ckeditor/ckeditor5/issues/1821)
+    // and therefore, we need to filter such logs until we've upgraded to
+    // CKEditor-5 completely.
+    browserLogs = browserLogs.filter(function(browserLog) {
```
Can we also file an issue for this and assign it to you? Else we might forget about this.

NishealJ replied:
Missed now updated PTAL!

NishealJ replied:
Done Added here #7224 Thanks!

nithusha21 commented on webpack.config.ts:
```diff
@@ -107,6 +110,9 @@ module.exports = {
   },
   plugins: [
+    new CKEditorWebpackPlugin({
+      language: 'pl'
```
What's this?

NishealJ replied:
Sorry, default config of CK-5 Updated to 'en' to avoid confusion.
this is used for i18n-translations as well

nithusha21 commented on webpack.prod.config.ts:
```diff
@@ -19,6 +19,8 @@
 var commonWebpackConfig = require('./webpack.config.ts');
```
Did you test this PR in both dev and prod environments?

NishealJ replied:
yes tested it prod env, build as expected Also tested via `start.sh --prod-env` to check the webpack build you have more suggestions?

NishealJ replied:
Done Added here #7224 Thanks!

nithusha21 commented on assets/constants.js:
```diff
@@ -15,6 +15,7 @@
 var constants = {
   // Whether to allow custom event reporting to Google Analytics.
   "CAN_SEND_ANALYTICS_EVENTS": false,
+  "IS_CURRENT_RTE_CKEDITOR_4" : true,
```
Add a blank line above it, and also add a comment describing the constant.

NishealJ replied:
Added PTAL!

nithusha21 reviewed (commented):
Hi @NishealJ, took a pass. Left some comments.

NishealJ committed:
Address comments

NishealJ committed:
update language

NishealJ committed:
Update codeowners and var names

NishealJ committed:
dev_mode

NishealJ commented:
Hi @nithusha21, Addressed all your comments PTAL Thanks!

apb7 commented:
Hi @NishealJ, one of the e2e tests is failing. PTAL! Also, whenever you've addressed all the comments posted by a reviewer, please re-assign the PR to him/her for another pass while leaving a top-level comment. Thanks!

NishealJ committed:
Fix-1

NishealJ committed:
Re-arrange CK

NishealJ committed:
Update few more directives

NishealJ committed:
Cross-chck

nithusha21 commented on core/templates/dev/head/pages/creator-dashboard-page/suggestion-modal-for-creator-view/suggestion-modal-for-creator-view.directive.html:
```diff
@@ -19,7 +19,12 @@ <h3>Review Suggestion</h3>
       </div>
       <div ng-show="suggestionEditorIsShown">
-        <ck-editor-4-rte ng-model="suggestionData.newSuggestionHtml"></ck-editor-4-rte>
+        <ck-editor-5-rte ng-if="!$ctrl.isDisabled() && !$ctrl.isCurrentRTECKEditor_4()"
```
Are you sure isDisabled() is a function in this directive, and are you sure it is needed here?

NishealJ replied:
Removed it was not needed + we have ng-show doing the same work.
this and the below one seems to be an independent editor I'll try replacing it will with the schema-editor we use at various places

nithusha21 commented on core/templates/dev/head/pages/exploration-player-page/templates/exploration-player-suggestion-modal.directive.html:
```diff
@@ -4,7 +4,12 @@ <h3>Suggest a Change</h3>
 <div class="modal-body protractor-test-exploration-suggestion-modal">
   <!-- ng-model needs to bind to a property of an object on the scope (the property cannot sit directly on the scope). -->
-  <ck-editor-4-rte ng-show="showEditor" ng-model="suggestionData.suggestionHtml"></ck-editor-4-rte>
+  <ck-editor-5-rte ng-if="!$ctrl.isDisabled() && !isCurrentRTECKEditor_4()" ng-show="showEditor"
```
Ditto, reg isDisabled()

NishealJ replied:
Done PTAL!

nithusha21 reviewed (commented):
Some follow up concerns. Other than these LGTM.

NishealJ committed:
Resolve issues

NishealJ commented:
Hi, @nithusha21 Addressed your comments PTAL Thanks!

nithusha21 reviewed (approved):
LGTM overall. Just confirming that you manually tested all of these locations right?

NishealJ commented:
> LGTM overall. Just confirming that you manually tested all of these locations right?

Hi @apb7 & @nithusha21,
yes I've tested again and added it in the testing doc(4th point) tested in all the needed places CK-4 is not affected in any means

apb7 reviewed (approved):
LGTM! Thanks, @NishealJ!

apb7 commented:
Hi @aks681, @DubeySandeep, @seanlip, @vinitamurthi and @vojtechjelinek, ptal as code owners. Thanks!

seanlip commented on assets/constants.js:
```diff
@@ -16,6 +16,9 @@ var constants = {
   "CAN_SEND_ANALYTICS_EVENTS": false,
 
+  // This specifies the current editor in use and used to switch between CK4 & CK5.
+  "IS_CURRENT_RTE_CKEDITOR_4" : true,
```
CURRENT_RTE_IS_CKEDITOR_4

(we usually start names with verbs only when they are methods/functions)

NishealJ replied:
Done Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div><div></div>' +
```
Why do you need the extra div at the start? Add a comment explaining why.

Also, structure this a bit more clearly:

```
 '<div>' +
 '  <div></div>' +
 '  <div ...></div>' +
 '</div>`
```

NishealJ replied:
Hi @seanlip, the extra div is needed for the toolbar the editor structure is something like this
```
<div> -> parent enclosure
   <div></div> -> toolbar
   <div></div> -> for editor textarea
</div>
```
I've copied the same template what we were using in CK4 because the structure is justifing and custom styleing can be applied to the toolbar through this structure if needed.

Below picture for reference :
<img width="654" alt="Screenshot 2019-07-27 at 11 54 51 PM" src="https://user-images.githubusercontent.com/41017303/61998189-f5c8bf80-b0c9-11e9-857f-626d7300b71b.png">

seanlip replied:
I see -- consider adding a comment inline to explain. It does help to understand that the first div is for the toolbar and the second is for the editor, but that's not clear from just looking at the template.

NishealJ replied:
Done as Suggested Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+      template: '<div><div></div>' +
+                '<div contenteditable="true" class="oppia-rte">' +
+                '</div></div>',
+      require: '?ngModel',
```
Why the "?" -- shouldn't all RTEs have a defined ngModel?

NishealJ replied:
yes, all RTEs need ngModel especially for plugins which have a modal for text input.
This is from CK-4 working because it was always able to find an ng-model controller.
let's make it `^^ngModel`? 
Also, do we have any other alternative to pass this?

seanlip replied:
I'm not 100% sure what you should do, but try and do something that actually raises an error if the condition is not met, and that is as strict and narrowly-defined as possible. So, either "no prefix", or "^", or "^^".

NishealJ replied:
Hi @seanlip, I've used `^` learned for [here](https://docs.angularjs.org/guide/directive), this will throw an error is its not found thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+                '</div></div>',
+      require: '?ngModel',
+
+      link: function(scope: ICustomScope, el, attr, ngModel) {
```
I *think* we use `elem` typically.

Is attr representing one or multiple attrs? If the latter, use attrs.

NishealJ replied:
Done updated it also attrs is corrected according to the docs [here](https://docs.angularjs.org/guide/directive) Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+        var _RICH_TEXT_COMPONENTS = RteHelperService.getRichTextComponents();
+        var names = [];
+        var icons = [];
+        var canUseFs = ContextService.getPageContext() ===
```
@aks681 has made some changes such that multiple pages with RTEs can use the file system (e.g. topics etc.). You should probably consult with him on this in case further changes are need to either your or his PRs.

NishealJ replied:
Hi @seanlip, I'll read more about it and update you here.
The changes made by @aks681 in fs seems to relatable here I'm looking into it

seanlip replied:
Sounds good. Consider actually talking to @aks681 too, as early in the process as possible -- he can give you feedback on your plan.

NishealJ replied:
I learned for @aks681 PR for new FS model and I've implemented the same here.
@aks681 PTAL Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+        var canUseFs = ContextService.getPageContext() ===
+          PAGE_CONTEXT.EXPLORATION_EDITOR;
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!((scope.uiConfig() &&
```
This is hard to understand, please make sure your code is readable and easy to follow "at a glance".

I suggest breaking this long clause out into separate variables, e.g. 

`var componentRequiresFsButFsCannotBeUsed = !canUseFs && componentDefn.requiresFs`

(perhaps with a shorter varname that indicates conceptually what you're checking) and use those.

(Though I am a bit confused about the one I just wrote above. If you can't use the filesystem and the component requires the filesystem then why are you including it?)

bansalnitish replied:
Hi @seanlip @NishealJ `canUseFs` shows whether file system can be used which is only possible when page context is editor mode. Further the component which requires fs is only image (We cannot add image while adding suggestions). So, a component should be disabled only when we cannot use fs and the component requires fs. In all other cases, we should add the component.That's the reason for the check before adding the component. 

Also the condition is a negation check `not (fs cannot be used and component requires fs)`. If the negation is true, component will be added. I think the confusion is due to the long condition in a single line. So, it is a good idea to split up as suggested by @seanlip!

NishealJ replied:
Thanks, @bansalnitish also regarding the image plugin how we upload the images we get from the custom image plugin which we are using now? seems like to migrate the image plugin we need to make an upload adapter (CK5 concept refer [here](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html)) do you have any other suggestions for this?

seanlip replied:
Hi @NishealJ -- I'm not sure I understand your question to @bansalnitish, but if the question is about how to implement the image plugin in CK5, I think you would need to find a solution (@bansalnitish may be able to offer suggestions, but I don't think he worked with CK5 much, so don't block on them).

For the way the current image plugin works, take a look at extensions/rte_components in Oppia, and look at the code for Image. It's implemented as a standard component to match all the other custom components, and ideally we do the same thing for CK5 -- i.e. have custom components that we maintain that all follow the same "format".

bansalnitish replied:
Hi @NishealJ, I am not sure if I understand your question clearly. We have added a custom plugin and widget for each rich text component (check `ck-editor-4-widgets.initializer.ts`). The widget on being clicked opens the customization modal for that component and then user can enter required details in the modal. Can you please  explain why do you need extra upload adapter for that?

NishealJ replied:
I was reading about it [here] (https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html) they have mentioned about using image-adapter. I've started the work for plugin migration and I'm finding the APIs which can initialize the rich text components as CK-widgets we had an API for the same in CKEditor 4.

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+        // a div with a CSS class.
+        var blockOverlayRule = ' div(oppia-rte-component-overlay);';
+        // Put all the rules together.
+        var extraAllowedContentRules = componentRule + inlineWrapperRule +
```
```
var extraAllowedContentRules = (
  componentRule + ... );
```

NishealJ replied:
Done PTAL Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+          scope.uiConfig().startupFocusEnabled !== undefined) {
+          startupFocusEnabled = scope.uiConfig().startupFocusEnabled;
+        }
+        // Initialize CKEditor.
```
Explain in the comment what the el[0], children[0] and children[1] refer to.

NishealJ replied:
Done Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -0,0 +1,93 @@
+        var ck = ClassicEditor.create(
+          <HTMLElement>(el[0].children[0].children[1]));
+
+        // A RegExp for matching rich text components.
```
This is never used? Neither is ck?

NishealJ replied:
`var ck` will be used in later to call the CK-5 APIs later that's why I've initialized it like that from the start Also, I think I should stack few more parts of CK migration otherwise this PR will be confusing

@apb7 can I do that?

seanlip replied:
I'm fine with it. I just wanted to make sure you didn't miss something.

NishealJ replied:
Hi @apb7 as discussed i've started the work this PR Thanks!

seanlip commented on core/templates/dev/head/components/forms/schema-based-editors/schema-based-html-editor.directive.ts:
```diff
@@ -35,6 +35,12 @@ oppia.directive('schemaBasedHtmlEditor', [
+      controller: ['$scope', 'IS_CURRENT_RTE_CKEDITOR_4',
+        function($scope, IS_CURRENT_RTE_CKEDITOR_4) {
+          var ctrl = this;
+          ctrl.isCurrentRTECKEditor_4 = function() {
```
isCurrentRteCKEditor4

vojtechjelinek replied:
Why you even need to have the function here? You can just do `isCurrentRTECKEditor_4 = IS_CURRENT_RTE_CKEDITOR_4`

NishealJ replied:
Done as suggested Thanks @seanlip & @vojtechjelinek

seanlip commented on core/templates/dev/head/pages/creator-dashboard-page/suggestion-modal-for-creator-view/suggestion-modal-for-creator-view.service.ts:
```diff
@@ -93,6 +93,10 @@ oppia.factory('SuggestionModalForCreatorDashboardService', [
               $scope.errorMessage = '';
             }
 
+            $scope.isCurrentRTECKEditor_4 = function() {
```
Ditto (naming) -- here and elsewhere

NishealJ replied:
Done PTAL Thanks!

seanlip commented on core/tests/protractor/accessibility.js:
```diff
@@ -49,6 +49,7 @@ describe('Cache Slugs', function() {
   it('should check that errors get logged for missing resources', function() {
     browser.get('/console_errors');
     var expectedErrors = [
+      'ckeditor-duplicated-modules',
```
Should you drop this if you're handling it more generally in general.js?

NishealJ replied:
Hi @seanlip, this matches the count of expected errors with the count of error summed from getting the library page. I found this test to be independent of general.js correct @apb7 ?

seanlip replied:
Ah, I think I misunderstood this. I think what you have is OK. Thanks!

aks681 reviewed (approved):
lgtm from codeowner's perspective.

DubeySandeep reviewed (approved):
LGTM for code owner files.

DubeySandeep commented on core/templates/dev/head/components/forms/schema-based-editors/schema-based-html-editor.directive.ts:
```diff
@@ -35,6 +35,12 @@ oppia.directive('schemaBasedHtmlEditor', [
         'schema-based-html-editor.directive.html'),
       controllerAs: '$ctrl',
-      controller: [function() {}]
+      controller: ['$scope', 'IS_CURRENT_RTE_CKEDITOR_4',
```
can we rename this such that it doesn't start with `IS`?

DubeySandeep commented on core/templates/dev/head/pages/exploration-editor-page/exploration-editor-page.controller.ts:
```diff
@@ -20,6 +20,7 @@
 // TODO(vojtechjelinek): this block of requires should be removed after we
 // introduce webpack for /extensions
 require('components/ck-editor-helpers/ck-editor-4-rte.directive.ts');
+require('components/ck-editor-helpers/ck-editor-5-rte.directive.ts');
```
If we can rename the file as `ck-editor-5-rte`, why can't we rename variable as 'CkEditor5Rte`? (is there any strict rule to have file names only with small caps?)

[Asking this just out of curiosity.]

DubeySandeep replied:
I see comment for the same over [here](https://github.com/oppia/oppia/pull/7184/files#r307957158). Marking resolved.

vojtechjelinek commented on webpack.dev.config.ts:
```diff
@@ -36,6 +38,29 @@ module.exports = {
   plugins: commonWebpackConfig.plugins,
   module: {
     rules: [{
+      test: /\.svg$/,
```
Please use the more specific regex here. (https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/advanced-setup.html#webpack-configuration) `/ckeditor5-[^/]+\/theme\/icons\/[^/]+\.svg$/`

NishealJ replied:
Done Thanks!

vojtechjelinek commented on webpack.dev.config.ts:
```diff
@@ -36,6 +38,29 @@ module.exports = {
+      use: ['raw-loader']
+    },
+    {
+      test: /\.css$/,
```
Please use the more specific regex here. (https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/advanced-setup.html#webpack-configuration) `/ckeditor5-[^/]+\/theme\/[\w-/]+\.css$/`

NishealJ replied:
Done Thanks!

vojtechjelinek commented on webpack.prod.config.ts:
```diff
@@ -36,6 +38,29 @@ module.exports = {
   plugins: commonWebpackConfig.plugins,
   module: {
     rules: [{
+      test: /\.svg$/,
```
ditto

NishealJ replied:
Done Thanks!

vojtechjelinek commented on webpack.prod.config.ts:
```diff
@@ -36,6 +38,29 @@ module.exports = {
+      use: ['raw-loader']
+    },
+    {
+      test: /\.css$/,
```
ditto

NishealJ replied:
Done Thanks!

vojtechjelinek reviewed (changes requested):
Thanks! Few comments.

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into ck-0-2

NishealJ committed:
Address Comments

NishealJ committed:
Address Comments-2

NishealJ committed:
Address comments-4

NishealJ committed:
Remove extra space

NishealJ commented:
Hi @apb7  @seanlip & @vojtechjelinek,
I've addressed all your comments PTAL Thanks! 
Also, @apb7 the linter is failing with an unexpected error seems like it is because of the regex added in webpack config how can I resolve that?

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -32,22 +32,28 @@ oppia.directive('ckEditor5Rte', [
+
         _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+         var componentRequiresFsButFsCannotBeUsed =
+        !canUseFs && componentDefn.requiresFs;
```
Indent this by 2 from the previous line. For clarity, use parens. I.e.:
```
        var componentRequiresFsButFsCannotBeUsed = (
          !canUseFs && componentDefn.requiresFs);
```

NishealJ replied:
Done PTAL Thanks!

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -32,22 +32,28 @@ oppia.directive('ckEditor5Rte', [
-                '<div contenteditable="true" class="oppia-rte">' +
-                '</div></div>',
-      require: '?ngModel',
+      template: '<div>' +
```
I don't see the comment I asked for before. You do have something lower down, but people won't read it because it's too far from this template. Comments should be near the things they're talking about.

So, can you put a comment above this template? E.g.:

```
      // The first div below is for the RTE toolbar, and the second div is for the ...
      template: '<div>' +
                '  <div></div>' +
                '  <div contenteditable="true" class="oppia-rte"></div>' +
                '</div>',
```

Also, in the template, indent the second and third lines by 2 spaces each (as I've done above). It doesn't change anything in the display to the end user, but it makes it easier to read over here.

NishealJ replied:
Done added a comment explaining the same.

seanlip commented on core/templates/dev/head/components/forms/schema-based-editors/schema-based-html-editor.directive.html:
```diff
@@ -1,11 +1,11 @@
-                 ui-config="$ctrl.uiConfig()">
-</ck-editor-4-rte>
-<ck-editor-5-rte ng-if="!$ctrl.isDisabled() && !$ctrl.isCurrentRTECKEditor_4()" ng-model="$ctrl.localValue"
+<ck-editor-5-rte ng-if="!$ctrl.isDisabled() && !$ctrl.isCurrentRteCKEditor4" ng-model="$ctrl.localValue"
```
Here and elsewhere -- I think I talked before about having functions start with a verb, and non-functions not start with a verb.

So, either $ctrl.isCurrentRteCKEditor4() (if it's a function), or $ctrl.currentRteIsCKEditor4 (if it's a boolean). But not $ctrl.isCurrentRteCKEditor4.

Please fix, here and in a bunch of places elsewhere.

NishealJ replied:
it was not a function so renamed it to `currentRteIsCKEditor4`

seanlip reviewed (changes requested):
Thanks, @NishealJ. A couple more things, otherwise LGTM.

seanlip commented:
> Also, @apb7 the linter is failing with an unexpected error seems like it is because of the regex added in webpack config how can I resolve that?

@NishealJ, it would be more helpful if you included a log (in your comment) of what the error is, as well as an explanation of what steps you've tried to figure out where it's coming from. Please do that in the future. Thanks.

NishealJ committed:
Address comments-1

NishealJ committed:
Address comments-2

NishealJ commented:
> @NishealJ, it would be more helpful if you included a log (in your comment) of what the error is, as well as an explanation of what steps you've tried to figure out where it's coming from. Please do that in the future. Thanks.
>
yes sure I'll keep to it, below is the error log @apb7 & @seanlip PTAL.
happens because of regex in webpack.dev.config.ts
```
Validating and parsing gulpfile.js file ...
Validating and parsing webpack.dev.config.ts file ...
Traceback (most recent call last):
  File "scripts/pre_commit_linter.py", line 2592, in <module>
    main()
  File "scripts/pre_commit_linter.py", line 2575, in main
    lint_checks_manager = LintChecksManager(all_filepaths, verbose_mode_enabled)
  File "scripts/pre_commit_linter.py", line 1119, in __init__
    self.parsed_js_and_ts_files = self._validate_and_parse_js_and_ts_files()
  File "scripts/pre_commit_linter.py", line 1163, in _validate_and_parse_js_and_ts_files
    raise Exception(e)
Exception: Line 40: Invalid regular expression
```

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts:
```diff
@@ -75,27 +80,26 @@ oppia.directive('ckEditor5Rte', [
         // Put all the rules together.
-        var extraAllowedContentRules = (
-          componentRule + inlineWrapperRule + blockWrapperRule + blockOverlayRule);
+        var extraAllowedContentRules = ( componentRule +
```
Should break after '('. Not sure why this got changed.

NishealJ replied:
The line exceeds 80 chars so I split it that way is it right?

seanlip replied:
If the contents of a paren spread out over more than one line, always break after the opening paren. You can add a further line break after one of the "+" signs if needed.

seanlip reviewed (changes requested):
Seems like the regex error is due to your changes, then.

Are you sure that the regexes you used are valid? When I enter them in https://regex101.com/ it shows an error.

NishealJ commented:
> Seems like the regex error is due to your changes, then.
> 
> Are you sure that the regexes you used are valid? When I enter them in https://regex101.com/ it shows an error.

I've added the regex based on the docs [https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/advanced-setup.html#webpack-configuration](https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/advanced-setup.html#webpack-configuration ) seems like it has an escape sequence error learning from [https://regex101.com/](https://regex101.com/ ) `ckeditor5-[^\/]+\/theme\/icons\/[^\/]+\.svg$` this must be correct

seanlip commented:
Also, all the e2e tests are failing...

seanlip commented:
Re the regex: ok, please make sure to test manually that everything is working and the right CKEditor assets are being added to the right places.

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

vojtechjelinek commented on core/templates/dev/head/pages/exploration-player-page/suggestion-modal-for-learner-local-view/suggestion-modal-for-exploration-player.service.ts:
```diff
@@ -64,6 +64,11 @@ oppia.factory('SuggestionModalForExplorationPlayerService', [
+
+            $scope.currentRteIsCKEditor4 = function() {
+              return CURRENT_RTE_IS_CKEDITOR_4;
+            };
```
Do you need function here?

NishealJ replied:
Hi @vojtechjelinek, no it was not needed I've updated it Thanks!

vojtechjelinek reviewed (approved):
LGTM! Just one nit

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into ck-0-2

NishealJ committed:
Address comments and regex

NishealJ committed:
Remove oppia module export

NishealJ commented:
Hi @seanlip & @apb7,
I still get this error `Exception: Line 40: Invalid regular expression` after validating the regex also.
I'm looking into it, rest all your comments have been addressed Thanks!

seanlip commented:
One suggestion: note that the linter code is available in the open source repo. So, you can try logging some stuff in the pre_commit_linter.py around the point where it fails, and do small isolated checks (e.g. change the main() function temporarily to see if you can run just the problematic part in isolation) to see if you can figure out what specifically is causing the failure. This will require reading and understanding some parts of the pre_commit_linter.py file.

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge develop into branch

NishealJ committed:
Update regex

NishealJ committed:
Fix lint in assets

NishealJ commented:
> One suggestion: note that the linter code is available in the open source repo. So, you can try logging some stuff in the pre_commit_linter.py around the point where it fails, and do small isolated checks (e.g. change the main() function temporarily to see if you can run just the problematic part in isolation) to see if you can figure out what specifically is causing the failure. This will require reading and understanding some parts of the pre_commit_linter.py file.

Thanks @seanlip, I debugged through the python regex library used in `scripts/pre_commit_linter.py` and was getting this error:
<img width="567" alt="Screenshot 2019-08-05 at 1 38 53 AM" src="https://user-images.githubusercontent.com/41017303/62428859-4d97a400-b724-11e9-8863-61c269357e55.png">
seems like the regex expression suggested by CKEditor had some issue, i've updated the regex expression learning it from [here](https://stackoverflow.com/questions/28539253/python-regex-bad-character-range) Thanks!

seanlip reviewed (approved):
Thanks! No further concerns.

@kevinlee12, I think this needs your review as codeowner, PTAL!

kevinlee12 reviewed (approved):
lgtm!

seanlip commented:
@NishealJ, feel free to ask for this to be merged once e2e tests pass. Thanks!

NishealJ commented:
I think I should update it with develop, getting an error
`The job exceeded the maximum time limit for jobs, and has been terminated.` Thanks!

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into develop

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into ck-0-2

NishealJ commented:
> @NishealJ, feel free to ask for this to be merged once e2e tests pass. Thanks!

Hi @seanlip, the e2e tests look good Thanks!

seanlip commented:
Merged. Thanks!
